core/includes/drupal.inc

```diff
@@ -2550,23 +2549,72 @@ function system_settings_form($form) {
  * forms, developers should create separate submit handlers for each form
  * to save the settings into various configuration files as needed.
  *
- * @deprecated since 1.0.0
  */
 function system_settings_form_submit($form, &$form_state) {
   watchdog_deprecated_function('drupal', __FUNCTION__);
   form_state_values_clean($form_state);
 
-  foreach ($form_state['values'] as $key => $value) {
-    if (is_array($value) && isset($form_state['values']['array_filter'])) {
-      $value = array_keys(array_filter($value));
+  if (!empty($form['#config'])) {
+    $configs = array();
+    backdrop_sort_form_values_by_config($form, $form_state, $configs, $form['#config']);
+    foreach ($configs as $bucket => $data) {
+      config_set_multiple($bucket, $data);
+    }
+  }
+  else {
+    watchdog('form', 'Backdrop 2.0 and later will require forms using system_settings_form() to set $form[\'#config\'].', array(), WATCHDOG_DEPRECATED);
+    foreach ($form_state['values'] as $key => $value) {
+      if (is_array($value) && isset($form_state['values']['array_filter'])) {
+        $value = array_keys(array_filter($value));
+      }
+      variable_set($key, $value);
     }
-    variable_set($key, $value);
   }
 
   backdrop_set_message(t('The configuration options have been saved.'));
 }
 
 /**
+ * Sorts the $form_state['values'] array into CMI buckets by $form['#config'].
+ *
+ * This function recursively searches the form to assign values from
+ * $form_state['values'] into specific CMI buckets.  The top of $form must
+ * contain a $form['#config'] entry with the name of the default CMI bucket
+ * to use.  If an element contains its own $form['#config'] entry, that CMI
+ * bucket name will be used for that element and any child elements in the
+ * form.
+ *
+ * @param  $form
+ *   An associative array containing the structure of a portion of the form.
+ * @param  &$form_state
+ *   A keyed array containing the current state of the form.
+ * @param  &$configs
+ *   A keyed array that will contain the form values sorted by CMI bucket when
+ *   the function returns.
+ * @param  $inherited_bucket
+ *   The name of the bucket to use if $form['#config'] is not set.
+ * @return none
+ */
+function backdrop_sort_form_values_by_config($form, &$form_state, &$configs, $inherited_bucket) {
+  foreach ($form as $key => $form_fragment) {
+    // Skip form metadata
+    if ($key[0] == '#') {
+      continue;
+    }
```

Comment:
The most typical approach for skipping form meta data like this is to use `element_children()` instead to get the list of keys. Though it's no more efficient, it would be consistent with the more common approach.